src/main/java/io/temporal/workflow/WorkflowInterceptor.java

```diff
@@ -19,89 +19,12 @@
 
 package io.temporal.workflow;
 
-import io.temporal.activity.ActivityOptions;
-import io.temporal.activity.LocalActivityOptions;
-import io.temporal.proto.common.WorkflowExecution;
-import io.temporal.workflow.Functions.Func;
-import java.lang.reflect.Type;
-import java.time.Duration;
-import java.util.Map;
-import java.util.Optional;
-import java.util.Random;
-import java.util.UUID;
-import java.util.function.BiPredicate;
-import java.util.function.Supplier;
-
+/**
+ * Intercepts workflow execution.
+ *
+ * <p>TODO(maxim): JavaDoc with sample
+ */
 public interface WorkflowInterceptor {
-
-  final class WorkflowResult<R> {
-
-    private final Promise<R> result;
-    private final Promise<WorkflowExecution> workflowExecution;
-
-    public WorkflowResult(Promise<R> result, Promise<WorkflowExecution> workflowExecution) {
-      this.result = result;
-      this.workflowExecution = workflowExecution;
-    }
-
-    public Promise<R> getResult() {
-      return result;
-    }
-
-    public Promise<WorkflowExecution> getWorkflowExecution() {
-      return workflowExecution;
-    }
-  }
-
-  <R> Promise<R> executeActivity(
-      String activityName,
-      Class<R> resultClass,
-      Type resultType,
-      Object[] args,
-      ActivityOptions options);
-
-  <R> Promise<R> executeLocalActivity(
-      String activityName,
-      Class<R> resultClass,
-      Type resultType,
-      Object[] args,
-      LocalActivityOptions options);
-
-  <R> WorkflowResult<R> executeChildWorkflow(
-      String workflowType,
-      Class<R> resultClass,
-      Type resultType,
-      Object[] args,
-      ChildWorkflowOptions options);
-
-  Random newRandom();
-
-  Promise<Void> signalExternalWorkflow(
-      WorkflowExecution execution, String signalName, Object[] args);
-
-  Promise<Void> cancelWorkflow(WorkflowExecution execution);
-
-  void sleep(Duration duration);
-
-  boolean await(Duration timeout, String reason, Supplier<Boolean> unblockCondition);
-
-  void await(String reason, Supplier<Boolean> unblockCondition);
-
-  Promise<Void> newTimer(Duration duration);
-
-  <R> R sideEffect(Class<R> resultClass, Type resultType, Func<R> func);
-
-  <R> R mutableSideEffect(
-      String id, Class<R> resultClass, Type resultType, BiPredicate<R, R> updated, Func<R> func);
-
-  int getVersion(String changeID, int minSupported, int maxSupported);
-
-  void continueAsNew(
-      Optional<String> workflowType, Optional<ContinueAsNewOptions> options, Object[] args);
-
-  void registerQuery(String queryType, Type[] argTypes, Functions.Func1<Object[], Object> callback);
-
-  UUID randomUUID();
-
-  void upsertSearchAttributes(Map<String, Object> searchAttributes);
+  WorkflowInvoker interceptExecuteWorkflow(
```

Comment:
Is that only going to be used by root context?  Why signal is missing here?

Reply:
WorkflowInterceptor.interceptExecuteWorkflow returns WorkflowInvoker which has:

- init() called when workflow class is instantiated
- execute() called when main workflow method is executed
- processSignal() called for each signal.